### PR TITLE
feat: java build changes

### PR DESCRIPTION
**Pull request by mndeveci**
Please don't merge this before merging this PR (https://github.com/aws/aws-lambda-builders/pull/318) and releasing new version of `aws-lambda-builders`.

#### Which issue(s) does this change fix?
Related to https://github.com/aws/aws-lambda-builders/pull/318

#### Why is this change necessary?
For using AWS Lambda Layers for Java runtimes and better dependency handling.

#### How does it address the issue?
- It builds Layers before Functions so that any Layer code can be used during Function build time
- It passes down 2 new parameters (Boolean flag to determine if it is building a Layer and list of experimental flags) to aws-lambda-builders so that changes in aws-lambda-builders can work

#### What side effects does this change have?
Any customer who is using `provided` scope for their `compile` or `runtime` dependency might be affected. Because of that we are making these changes behind a feature flag before making this behavior as default.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/build/build_context.py [hunk 1]
```diff
@@ -508,3 +509,24 @@ def _is_layer_buildable(layer: LayerVersion):
...
             return False
         return True
+
+    _JAVA_BUILD_WARNING_MESSAGE = (
+        "Test the latest build changes for Java runtime 'SAM_CLI_BETA_MAVEN_SCOPE_AND_LAYER=1 sam build'. "
+        "These changes will replace the existing flow on 1st of April 2022. "
+        "Check __issue_line__ for more information."
+    )
+
+    def _check_java_warning(self):
```
Comment (mildaniel) [inline samcli/commands/build/build_context.py:519]: nit: I know we don't add type hints and docstrings to older files like this, but should we start for consistency going forward?

**Review by mildaniel — APPROVED** [review]

File: samcli/commands/build/build_context.py (hunk 1 above)
Comment (mndeveci) [reply samcli/commands/build/build_context.py:519]: Good point, I will update the PR

File: samcli/commands/build/build_context.py [hunk 2]
```diff
@@ -508,3 +509,29 @@ def _is_layer_buildable(layer: LayerVersion):
             LOG.debug("Skip building pre-built layer: %s", layer.full_path)
             return False
         return True
+
+    _JAVA_BUILD_WARNING_MESSAGE = (
+        "Test the latest build changes for Java runtime 'SAM_CLI_BETA_MAVEN_SCOPE_AND_LAYER=1 sam build'. "
+        "These changes will replace the existing flow on 1st of April 2022. "
+        "Check __issue_line__ for more information."
```
Comment (CoshUS) [inline samcli/commands/build/build_context.py:516]: What's the `__issue_line__` referring to?

**Review by CoshUS — COMMENTED** [review]
Thanks for the PR. Added one question.

File: samcli/commands/build/build_context.py (hunk 2 above)
Comment (mndeveci) [reply samcli/commands/build/build_context.py:516]: We will update this warning message once we have an announcement GH issue ready in SAM CLI. This is just a placeholder for that

**Review by CoshUS — APPROVED** [review]

**Pull request merged**